src/async_context_frame.cc

```diff
@@ -0,0 +1,270 @@
+#include "async_context_frame.h"  // NOLINT(build/include_inline)
+
+#include "env-inl.h"
+#include "node_errors.h"
+#include "node_external_reference.h"
+#include "tracing/traced_value.h"
+#include "util-inl.h"
+
+#include "debug_utils-inl.h"
+
+#include "v8.h"
+
+using v8::Context;
+using v8::FunctionCallbackInfo;
+using v8::FunctionTemplate;
+using v8::Isolate;
+using v8::Local;
+using v8::Object;
+using v8::Value;
+
+#if defined(NODE_USE_NATIVE_ALS) && NODE_USE_NATIVE_ALS
+
+namespace node {
+
+//
+// Scope helper
+//
+AsyncContextFrame::Scope::Scope(Isolate* isolate, Local<Value> object)
+    : isolate_(isolate) {
+  auto prior = AsyncContextFrame::exchange(isolate, object);
+  prior_.Reset(isolate, prior);
+}
+
+AsyncContextFrame::Scope::Scope(Isolate* isolate, AsyncContextFrame* frame)
+    : Scope(isolate, frame->object()) {}
+
+AsyncContextFrame::Scope::~Scope() {
+  auto value = prior_.Get(isolate_);
+  AsyncContextFrame::exchange(isolate_, value);
+}
+
+//
+// Constructor
+//
+AsyncContextFrame::AsyncContextFrame(Environment* env,
+                                     Local<Object> obj,
+                                     Local<Object> current,
+                                     Local<Value> key,
+                                     Local<Value> value)
+    : BaseObject(env, obj),
+      parent_(env->isolate(), current),
+      key_(env->isolate(), key),
+      value_(env->isolate(), value),
+      enabled_(true) {
+  key_.SetWeak();
+}
+
+Local<Value> AsyncContextFrame::current(Isolate* isolate) {
+  return isolate
+      ->GetEnteredOrMicrotaskContext()
+      ->GetContinuationPreservedEmbedderData();
+}
+
+// NOTE: It's generally recommended to use AsyncContextFrame::Scope
+// but sometimes (such as enterWith) a direct exchange is needed.
+Local<Value> AsyncContextFrame::exchange(Isolate* isolate,
+                                                Local<Value> value) {
+  auto prior = current(isolate);
+  isolate
+      ->GetEnteredOrMicrotaskContext()
+      ->SetContinuationPreservedEmbedderData(value);
+  return prior;
+}
+
+Local<Value> AsyncContextFrame::disable(Isolate* isolate, Local<Value> key) {
+  if (key_ == key) {
+    enabled_ = false;
+    return v8::True(isolate);
+  }
+
+  auto parent = parent_.Get(isolate);
+  Environment* env = Environment::GetCurrent(isolate);
+  if (parent.IsEmpty() || !AsyncContextFrame::HasInstance(env, parent)) {
+    return v8::False(isolate);
+  }
+
+  return Unwrap<AsyncContextFrame>(parent)->disable(isolate, key);
+}
```

Comment:
@anonrig Do you think it's possible to add FastAPI support to these calls?

The return can be a boolean and the `key` (from what I've read) is an instance of `AsyncContextFrame`, which from what I've read in [v8-fast-api-calls.h](https://github.com/nodejs/node/blob/main/deps/v8/include/v8-fast-api-calls.h#L53-L91), can be thought of as an embedding type.

My only concern is related to `AsyncContextFrame::HasInstance(env, parent)` which requires `env`, [on timers.cc](https://github.com/nodejs/node/blob/3af65855c5fa2626fba98c2cf0cd70bd52256800/src/timers.cc#L69) they can access env, but it's considered the BindingData.

Reply:
The disable is so rarely called I don't think it's worth the effort. If there was a way to do that on `current` or `exchange` that would be _super_ valuable, but I don't see how as that needs to interact with heap objects so violates the no-gc rule, as far as I understand. 🤔

Reply:
Maybe using `BindingData` we can add support for `current` and `exchange` following the same pattern as `timers.h` but I'm not sure, let's wait for Yagiz's answer and see if it's worth exploring.

Reply:
Function with a string parameter and a return value of boolean is supported by the fast apis.

Reply:
I looked briefly but I believe most of these functions can be implemented in fast apis (which is a once in a lifetime opportunity) 😅

Reply:
The key is not a string, it's an instance of `AsyncLocalStorage`.

Reply:
I don't think that is a problem, in their own example they use [v8::Local<v8::Object>](https://github.com/nodejs/node/blob/main/deps/v8/include/v8-fast-api-calls.h#L73) and then cast to `CustomEmbedderType`

Reply:
From the supported types section it seems to indicate Value types would not be a supported return value so neither `current` nor `exchange` could work. A separate `set` could _maybe_ work, but would require a complete restructuring as it would depend on `AsyncLocalStorage` being an embedder type, so I would need to reimplement it in C++, if I'm understanding correctly.

Reply:
Exactly, `current` cannot be converted to fast-api but for `exchange` you can create an `exchangeNoReturn` just to use this fast-api because sometimes you don't use return from `exchange`, in such cases you can speed it up a bit, [like this one.](https://github.com/nodejs/node/pull/48528/files/de98e97199c7e7c9132fab92b5846f23b83b1acc#diff-41813f5e6561d74a1162a1a326c914dd05d298c44c90653726d3997bdc1f9df1R506)